app/src/main/kotlin/io/homeassistant/companion/android/widgets/todo/TodoWidgetState.kt

```diff
@@ -0,0 +1,122 @@
+                uid = todoItem.uid,
+                name = todoItem.summary ?: "",
+                done = todoItem.status == COMPLETED_STATUS,
+            )
+        }
+    }
+}
+
+internal sealed interface TodoState {
+    val backgroundType: WidgetBackgroundType
+    val textColor: String?
+
+    companion object {
+        @Composable
+        fun TodoState.getColors(): ColorProviders {
+            return when (backgroundType) {
+                WidgetBackgroundType.DYNAMICCOLOR -> GlanceTheme.colors
+                WidgetBackgroundType.DAYNIGHT -> HomeAssistantGlanceTheme.colors
+                WidgetBackgroundType.TRANSPARENT -> ColorProviders(
+                    glanceHaLightColors
+                        .copy(background = Color.Transparent, onSurface = Color(textColor?.toColorInt() ?: glanceHaLightColors.onSurface.toArgb())),
+                )
+            }
+        }
+    }
+}
+
+internal data object LoadingTodoState : TodoState {
+    override val backgroundType: WidgetBackgroundType = WidgetBackgroundType.DAYNIGHT
+    override val textColor: String? = null
+}
+
+internal data object EmptyTodoState : TodoState {
+    override val backgroundType: WidgetBackgroundType
+        get() = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+            WidgetBackgroundType.DYNAMICCOLOR
+        } else {
+            WidgetBackgroundType.DAYNIGHT
+        }
+    override val textColor: String? = null
+}
+
+internal data class TodoStateWithData(
+    override val backgroundType: WidgetBackgroundType,
+    override val textColor: String?,
+    val serverId: Int,
+    val listEntityId: String,
+    val listName: String?,
+    val todoItems: List<TodoItemState>,
+    val outOfSync: Boolean,
+    val showComplete: Boolean,
+) : TodoState {
+
+    fun hasDisplayableItems(): Boolean {
+        return if (showComplete) {
+            todoItems.isNotEmpty()
+        } else {
+            todoItems.any { !it.done }
+        }
+    }
+
+    companion object {
+        /**
+         * Create a complete [TodoStateWithData] from the DB and from the server. Set the flag [outOfSync] to false, since the data
+         * should just come from the server.
+         */
+        fun from(todoEntity: TodoWidgetEntity, entity: Entity<*>, todos: List<TodoWidgetEntity.TodoItem>): TodoStateWithData {
+            return TodoStateWithData(
+                backgroundType = todoEntity.backgroundType,
+                textColor = todoEntity.textColor,
+                serverId = todoEntity.serverId,
+                listEntityId = entity.entityId,
+                listName = entity.friendlyName,
+                todoItems = todos.map(TodoItemState::from),
+                outOfSync = false,
+                showComplete = todoEntity.showCompleted,
+            )
+        }
+
+        /**
+         * Since we only have the DB entity it means that we are out of sync. So when using this function it will
+         * set the flag [outOfSync] to true.
```

Comment:
I think it's nicer to align the descriptions.

```suggestion
         * Create a [TodoStateWithData] with data only from the DB. Set the flag [outOfSync] to true, since the data
         * doesn't have an updated state from the server.
```